testsuite/tests/lib-obj/obj_dup_closures.ml

```diff
@@ -0,0 +1,173 @@
+  in
+  let[@opaque] c2_1arg () =
+    (* An unboxed environment plus a scannable environment *)
+    let (_i : int) = Sys.opaque_identity (String.length x) in
+    i64_2
+  in
+  let[@opaque] c3_1arg () =
+    (* Only a scannable environment *)
+    let (_i : int) = Sys.opaque_identity (String.length x) in
+    100
+  in
+  (* Three-word function slot (i.e. more than one argument) cases *)
+  let[@opaque] c1_2arg () () =
+    i64_3
+  in
+  let[@opaque] c2_2arg () () =
+    let (_i : int) = Sys.opaque_identity (String.length x) in
+    i64_4
+  in
+  let[@opaque] c3_2arg () () =
+    let (_i : int) = Sys.opaque_identity (String.length x) in
+    200
+  in
+  (* Cases to exercise [Infix_tag] logic *)
+  let[@opaque] rec rec_c1_1arg () =
+    i64_1
+  and[@opaque] rec_c2_1arg () =
+    let (_i : int) = Sys.opaque_identity (String.length x) in
+    i64_2
+  and[@opaque] rec_c3_1arg () =
+    let (_i : int) = Sys.opaque_identity (String.length x) in
+    300
+  and[@opaque] rec_c1_2arg () () =
+    let (_ : int64#) = Sys.opaque_identity i64_3 in
+    rec_c1_1arg ()
+  and[@opaque] rec_c2_2arg () () =
+    let (_i : int) = Sys.opaque_identity (String.length x) in
+    let (_ : int64#) = Sys.opaque_identity i64_4 in
+    rec_c2_1arg ()
+  and[@opaque] rec_c3_2arg () () =
+    let (_i : int) = Sys.opaque_identity (String.length x) in
+    rec_c3_1arg ()
+  in
+  c1_1arg, c2_1arg, c3_1arg,
+  c1_2arg, c2_2arg, c3_2arg,
+  rec_c1_1arg, rec_c2_1arg, rec_c3_1arg,
+  rec_c1_2arg, rec_c2_2arg, rec_c3_2arg
+
+let check
+    (i64_1 : int64#) (i64_2 : int64#) (i64_3 : int64#) (i64_4 : int64#)
+    (x : string)
+    (c1_1arg, c2_1arg, c3_1arg,
+    c1_2arg, c2_2arg, c3_2arg,
+    rec_c1_1arg, rec_c2_1arg, rec_c3_1arg,
+    rec_c1_2arg, rec_c2_2arg, rec_c3_2arg) =
+  assert (Int64_u.equal (c1_1arg ()) i64_1);
+  assert (Int64_u.equal (c2_1arg ()) i64_2);
+  assert (Int.equal (c3_1arg ()) 100);
+  assert (Int64_u.equal (c1_2arg () ()) i64_3);
+  assert (Int64_u.equal (c2_2arg () ()) i64_4);
+  assert (Int.equal (c3_2arg () ()) 200);
+  assert (Int64_u.equal (rec_c1_1arg ()) i64_1);
+  assert (Int64_u.equal (rec_c2_1arg ()) i64_2);
+  assert (Int.equal (rec_c3_1arg ()) 300);
+  assert (Int64_u.equal (rec_c1_2arg () ()) i64_1);
+  assert (Int64_u.equal (rec_c2_2arg () ()) i64_2);
+  assert (Int.equal (rec_c3_2arg () ()) 300)
+
+let check_one () =
+  let i64_1 = rand_near_minor_heap () in
+  let i64_2 = rand_near_minor_heap () in
+  let i64_3 = rand_near_minor_heap () in
+  let i64_4 = rand_near_minor_heap () in
+  let x = rand_string () in
+  let closures :
+      (unit -> int64#) * (unit -> int64#) * (unit -> int) *
+      (unit -> unit -> int64#) * (unit -> unit -> int64#) *
+      (unit -> unit -> int) * (unit -> int64#) * (unit -> int64#) *
+      (unit -> int) * (unit -> unit -> int64#) * (unit -> unit -> int64#) *
+      (unit -> unit -> int) =
+    make_closures i64_1 i64_2 i64_3 i64_4 x
+    |> Obj.repr  (* this includes opaqueness *)
+    |> Obj.dup
```

Comment:
This will not copy the closures, only the block containing them I think?

Reply:
Whoops, fixed